src/librustc/ty/sty.rs

```diff
@@ -2152,20 +2152,19 @@ impl<'tcx> Const<'tcx> {
     }
 
     #[inline]
-    pub fn assert_bits(
-        &self,
-        tcx: TyCtxt<'_, '_, '_>,
-        ty: ParamEnvAnd<'tcx, Ty<'tcx>>,
-    ) -> Option<u128> {
+    pub fn assert_bits(&self, ty: ParamEnvAnd<'tcx, Ty<'tcx>>) -> Option<u128> {
         assert_eq!(self.ty, ty.value);
-        let ty = tcx.lift_to_global(&ty).unwrap();
-        let size = tcx.layout_of(ty).ok()?.size;
-        self.val.try_to_bits(size)
+        match self.val.try_to_scalar()? {
+            Scalar::Bits { bits, .. } => {
+                Some(bits)
```

Comment:
This is fishy, we shouldn't usually return `bits` without checking that the size is correct.

Reply:
The idea is that they should already be correct because the type of the constant is the same. Sanity checks would have bailed long before if the number of bits didn't match the type.

Reply:
Are constants expected to have sizes that don't match their type's size? If so, the premise to this PR is broken. If not, it would seem more sensible to me to assert that the sizes match when the constant is created, rather than on each time we read from it.

Reply:
> Are constants expected to have sizes that don't match their type's size?

No, that's why there are assertions in place that check the size. If we happened to get the size wrong, that's a bug. These assertions just make sure that we don't screw up (as we have done before, this is not just hypothetical, it's easy to get wrong in some places, but less so nowadays).

> If not, it would seem more sensible to me to assert that the sizes match when the constant is created, rather than on each time we read from it.

I believe that we are doing this now, not directly when creating the `ty::Const`, but during `const_eval`.

Reply:
TBH I am not very happy about removing a sanity check that costs no measurable performance and caught real bugs.  So I'd r-.

But if @oli-obk vetoes me on this, that's fine for me.  Just please add a comment stating very explicitly that we are deliberately omitting a sanity check here.

Reply:
Oh, right, the assertion would have triggered then anyway. I've just been confused, because @RalfJung's comment sounded to me as if the type equality assertion is not enough, and the sizes could be mismatched regardless.

Reply:
> the type equality assertion is not enough, and the sizes could be mismatched regardless.

It is not enough and there could be a mismatch, pointing to a bug elsewhere. I am not sure if all `Const` constructors have this check.